### PR TITLE
Feature/update program state

### PR DESCRIPTION
**Pull request by Mee-gu**
remove BindGroup, add ProgramState

### CODE REVIEW

File: cocos/renderer/backend/ProgramCache.cpp [hunk 1]
```diff
@@ -0,0 +1,130 @@
+#include "ProgramCache.h"
+#include "Device.h"
+#include "ShaderModule.h"
+#include "renderer/ccShaders.h"
+
+CC_BACKEND_BEGIN
+
+std::unordered_map<std::size_t, backend::Program*>  ProgramCache::_cachedPrograms;
+static ProgramCache *_sharedProgramCache = nullptr;
```
Comment (minggo) [inline cocos/renderer/backend/ProgramCache.cpp]: please use static member variable

File: cocos/renderer/CCProgramState.cpp [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+void ProgramState::setVertexTexture(int location, uint32_t slot, backend::Texture* texture)
+{
+    setTexture(location, slot, texture, _vertexTextureInfos);
+}
+
+void ProgramState::setFragmentTexture(int location, uint32_t slot, backend::Texture* texture)
+{
+    if(location < 0)
+        printf("texture location = %d", location);
```
Comment (minggo) [inline cocos/renderer/CCProgramState.cpp]: please use CCLOG instead

File: cocos/renderer/CCShaderCache.cpp [hunk 3]
```diff
@@ -51,11 +54,13 @@ backend::ShaderModule* ShaderCache::newFragmentShaderModule(const std::string& s
...
         return iter->second;
     
-    auto shader = backend::Device::getInstance()->createShaderModule(backend::ShaderStage::FRAGMENT, shaderSource);
-    CC_SAFE_RETAIN(shader);
-    ShaderCache::_cachedFragmentShaders.emplace(key, shader);
+//    auto shader = backend::Device::getInstance()->createShaderModule(backend::ShaderStage::FRAGMENT, shaderSource);
+//    CC_SAFE_RETAIN(shader);
+//    ShaderCache::_cachedFragmentShaders.emplace(key, shader);
+    
+    printf("----shaderCache-------------");
```
Comment (minggo) [inline cocos/renderer/CCShaderCache.cpp]: please remove unneeded codes

File: cocos/renderer/backend/metal/CommandBufferMTL.h [hunk 4]
```diff
@@ -33,9 +35,9 @@ class CommandBufferMTL : public CommandBuffer
 private:
     void prepareDrawing() const;
     void setTextures() const;
-    void doSetTextures(const std::vector<std::string>& textures, bool isVertex) const;
+    void doSetTextures(bool isVertex) const;
     void setUniformBuffer() const;
-    unsigned int fillUniformBuffer(uint8_t* buffer, const std::vector<std::string>& uniforms) const;
+    uint32_t fillUniformBuffer(uint8_t* buffer, const std::vector<UniformBuffer>& unifornInfo) const;
```
Comment (minggo) [inline cocos/renderer/backend/metal/CommandBufferMTL.h]: please use unsigned int instead

File: cocos/renderer/backend/metal/CommandBufferMTL.mm [hunk 5]
```diff
@@ -209,11 +208,10 @@ MTLCullMode toMTLCullMode(CullMode mode)
...
 }
 
-void CommandBufferMTL::setBindGroup(BindGroup* bindGroup)
+void CommandBufferMTL::setProgramState(ProgramState* programState)
 {
-    CC_SAFE_RETAIN(bindGroup);
-    CC_SAFE_RELEASE(_bindGroup);
-    _bindGroup = bindGroup;
+    CC_SAFE_RETAIN(programState);
+    _programState = programState;
```
Comment (minggo) [inline cocos/renderer/backend/metal/CommandBufferMTL.mm:215]: potential memory leak

File: cocos/renderer/backend/opengl/ProgramGL.h [hunk 6]
```diff
@@ -5,57 +5,59 @@
...
 CC_BACKEND_BEGIN
 
 class ShaderModuleGL;
 
 struct AttributeInfo
 {
-    unsigned int location = 0;
-    unsigned int size = 0;
+    uint32_t location = 0;
+    uint32_t size = 0;
```
Comment (minggo) [inline cocos/renderer/backend/opengl/ProgramGL.h]: please use unsigned int instead

File: cocos/renderer/CCProgramState.cpp [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+    _fragmentTextureInfos.clear();
+}
+
+void ProgramState::createVertexUniformBuffer()
+{
+    const auto& vertexUniformInfos = _program->getVertexUniformInfos();
+    for(const auto& uniformInfo : vertexUniformInfos)
+    {
+        if(uniformInfo.second.bufferSize)
+            _vertexUniformInfos.at(uniformInfo.second.location) = uniformInfo.second;
```
Comment (minggo) [inline cocos/renderer/CCProgramState.cpp]: please use `[]` instead of `at` to improve performance.

File: cocos/renderer/CCProgramState.cpp [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+}
+
+void ProgramState::setVertexUniform(int location, const void* data, uint32_t size)
+{
+    if(location < 0)
+        return;
+    
+    assert(size <= _vertexUniformInfos.at(location).uniformInfo.bufferSize);
+    memcpy(_vertexUniformInfos.at(location).data, data, size);
+    _vertexUniformInfos.at(location).dirty = true;
```
Comment (minggo) [inline cocos/renderer/CCProgramState.cpp]: please use `[]` instead of `at` to improve performance.

**Comment by Mee-gu** [comment]
update review

File: cocos/2d/CCGrid.h [hunk 9]
```diff
@@ -155,6 +155,11 @@ class CC_DLL GridBase : public Ref
     Texture2D* _oldColorAttachment = nullptr;
     Texture2D* _oldDepthAttachment = nullptr;
     Texture2D* _oldStencilAttachment = nullptr;
+    
+    backend::UniformLocation _mvpMatrixLocation;
+    backend::UniformLocation _textureLocation;
```
Comment (minggo) [inline cocos/2d/CCGrid.h]: should initialize it

File: cocos/2d/CCLabel.cpp [hunk 10]
```diff
@@ -1523,9 +1529,11 @@ void Label::updateEffectUniforms(TextureAtlas* textureAtlas, Renderer *renderer,
         updateBuffer(textureAtlas, _customCommandShadow);
         auto& pipelineShadow = _customCommandShadow.getPipelineDescriptor();
         pipelineShadow = pipelineDescriptor;
+        pipelineShadow.programState = pipelineDescriptor.programState;
+        CC_SAFE_RETAIN(pipelineShadow.programState);
```
Comment (minggo) [inline cocos/2d/CCLabel.cpp]: don't have to retain

File: cocos/2d/CCLabel.cpp [hunk 11]
```diff
@@ -1554,17 +1562,19 @@ void Label::updateEffectUniforms(TextureAtlas* textureAtlas, Renderer *renderer,
                     updateBuffer(textureAtlas, _customCommandOutLine);
                     auto& pipelineOutline = _customCommandOutLine.getPipelineDescriptor();
                     pipelineOutline = pipelineDescriptor;
-                    pipelineOutline.bindGroup.setUniform("u_effectColor", &effectColor, sizeof(Vec4));
-                    pipelineOutline.bindGroup.setUniform("u_effectType", &effectType, sizeof(effectType));
+                    pipelineOutline.programState = pipelineDescriptor.programState;
+                    CC_SAFE_RETAIN(pipelineOutline.programState);
```
Comment (minggo) [inline cocos/2d/CCLabel.cpp]: don't have to retain it

File: cocos/2d/CCLabel.cpp [hunk 12]
```diff
@@ -1659,12 +1669,14 @@ void Label::draw(Renderer *renderer, const Mat4 &transform, uint32_t flags)
             auto texture = textureAtlas->getTexture();
             auto& pipelineQuad = _quadCommand.getPipelineDescriptor();
             pipelineQuad = pipelineDescriptor;
-            pipelineQuad.bindGroup.setUniform("u_MVPMatrix", matrixProjection.m, sizeof(matrixProjection.m));
-            pipelineQuad.bindGroup.setTexture("u_texture", 0, texture->getBackendTexture());
+            pipelineQuad.programState = pipelineDescriptor.programState;
+            CC_SAFE_RETAIN(pipelineQuad.programState);
```
Comment (minggo) [inline cocos/2d/CCLabel.cpp]: don't have to retain it

File: cocos/2d/CCSprite.cpp [hunk 13]
```diff
@@ -1107,8 +1120,10 @@ void Sprite::draw(Renderer *renderer, const Mat4 &transform, uint32_t flags)
 #endif
     {
         const auto& projectionMat = Director::getInstance()->getMatrix(MATRIX_STACK_TYPE::MATRIX_STACK_PROJECTION);
-        auto& bindGroup = _trianglesCommand.getPipelineDescriptor().bindGroup;
-        bindGroup.setUniform("u_MVPMatrix", projectionMat.m, sizeof(projectionMat.m));
+        auto programState = _trianglesCommand.getPipelineDescriptor().programState;
+        programState->setUniform(_mvpMatrixLocation, projectionMat.m, sizeof(projectionMat.m));
+        if(_programStateDirty)
+            updateProgramState();
```
Comment (minggo) [inline cocos/2d/CCSprite.cpp]: should update texture in `updateShaders ` and `setTexture `.

File: cocos/2d/CCSprite.h [hunk 14]
```diff
@@ -633,6 +633,8 @@ CC_CONSTRUCTOR_ACCESS :
 
     virtual void flipX();
     virtual void flipY();
+    
+    void updateProgramState();
```
Comment (minggo) [inline cocos/2d/CCSprite.h]: it is better to be virtual

File: cocos/2d/CCTMXLayer.cpp [hunk 15]
```diff
@@ -248,9 +250,14 @@ void TMXLayer::parseInternalProperties()
             auto alphaFuncVal = getProperty("cc_alpha_func");
             float alphaFuncValue = alphaFuncVal.asFloat();
 
+            cocos2d::log("TODO in %s %s %d", __FILE__, __FUNCTION__, __LINE__);
```
Comment (minggo) [inline cocos/2d/CCTMXLayer.cpp]: don't have to be todo

File: cocos/renderer/backend/Device.h [hunk 16]
```diff
@@ -33,15 +38,19 @@ class Device : public cocos2d::Ref
...
     // Create a auto released blend state.
     virtual BlendState* createBlendState(const BlendDescriptor& descriptor) = 0;
     // Create a render pipeline, not auto released.
     virtual RenderPipeline* newRenderPipeline(const RenderPipelineDescriptor& descriptor) = 0;
     
+protected:
+    // Create a auto released shader module.
+    virtual ShaderModule* createShaderModule(ShaderStage stage, const std::string& source) = 0;
+    // Create a auto released program.
+    virtual Program* createProgram(const std::string& vertexShader, const std::string& fragmentShader) = 0;
```
Comment (minggo) [inline cocos/renderer/backend/Device.h]: createProgram -> newProgram

**Comment by Mee-gu** [comment]
update review

**Pull request merged**